src/lib.rs

```diff
@@ -120,43 +120,32 @@ pub enum ClangFormatError {
 pub fn clang_format_with_style(
     input: &str,
     style: &ClangFormatStyle,
-) -> Result<String, ClangFormatError> {
+) -> Result<String, impl std::error::Error> {
     // Create and try to spawn the command with the specified style
     let clang_binary = env::var("CLANG_FORMAT_BINARY").unwrap_or("clang-format".to_string());
-    if let Ok(mut child) = Command::new(clang_binary.as_str())
+    let mut child = Command::new(clang_binary.as_str())
         .arg(format!("--style={}", style.as_str()))
         .stdin(Stdio::piped())
         .stdout(Stdio::piped())
-        .spawn()
+        .spawn()?;
+
+    // Write the input to stdin
+    //
+    // Note we place inside a scope to ensure that stdin is closed
     {
-        // Try to take the stdin pipe
-        if let Some(mut stdin) = child.stdin.take() {
-            // Write the input to the stdin
-            if write!(stdin, "{}", input).is_err() {
-                return Err(ClangFormatError::StdInWriteFailure);
-            }
-        } else {
-            return Err(ClangFormatError::StdInFailure);
-        }
+        let mut stdin = child.stdin.take().expect("no stdin handle");
+        write!(stdin, "{}", input)?;
+    }
 
-        // Wait for the output
-        //
-        // Note this cannot be inside the stdin block, as stdin is only closed
-        // when it goes out of scope
-        if let Ok(output) = child.wait_with_output() {
-            // Parse the output into a String
-            //
-            // TODO: do we need to check stderr or exitcode?
-            if let Ok(stdout) = String::from_utf8(output.stdout) {
-                Ok(stdout)
-            } else {
-                Err(ClangFormatError::Utf8FormatError)
-            }
-        } else {
-            Err(ClangFormatError::WaitFailure)
-        }
+    // Wait for the output and parse it
+    let output = child.wait_with_output()?;
+    // TODO: use exit_ok() once it is a stable feature
+    // https://doc.rust-lang.org/stable/std/process/struct.ExitStatus.html#method.exit_ok
+    // https://github.com/rust-lang/rust/issues/84908
+    if output.status.success() {
+        String::from_utf8(output.stdout).map_err(ClangFormatError::from)
```

Comment:
```suggestion
        Ok(String::from_utf8(output.stdout)?)
```

Hm, this is a very minor thing, but the current code I think reads a bit as if there is never "Ok" returned.
That's probably just me being used to seeing `Ok(...)` at the end of a function though.
The map_err approach is more functional as well, which is also nice.

Reply:
I think the Ok thing is slightly more readable what is happening and avoids the awkward `ClangFormatError::from`, so i'll change it.